File: src/dataStructures/graphs/validTree.rb
# @param {Integer} n
# @param {Integer[][]} edges
# @return {Boolean}
def valid_tree(n, edges)
    dsu = DSU.new(n)
    components = n
    edges.each do |edge|
        answer = dsu.union(edge[0], edge[1])
        return false if answer == 0
        components -= answer
    end
    components == 1 ? true : false
end

class DSU
    attr_accessor :parent, :rank

    def initialize(n)
        @parent = Array.new(n) {|i| i}
        @rank = Array.new(n) {1}
    end

    def find(x)
        if parent[x] != x
            parent[x] = find(parent[x])
        end
        parent[x]
    end

    def union(x, y)
        x = find(x)
        y = find(y)
        return 0 if x == y
        if rank[x] > rank[y]
            parent[y] = x
        elsif rank[x] < rank[y]
            parent[x] = y
        else
            parent[y] = x
            rank[x] += 1
        end
        1
    end
end

File: src/dataStructures/graphs/validTree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {boolean}
 */
 var validTree = function(n, edges) {
    let components = n;
    const dsu = new DSU(n);
    for(let edge of edges) {
        const union = dsu.uniion(edge[0], edge[1]);
        if(!union) return false;
        components -= union;
    }
    return components === 1 ? true : false;
};

// Disjoint set union data structure
class DSU {
    constructor(n) {
        this.parent = new Array(n).fill(0).map((el, index) => index);
        this.rank = new Array(n).fill(1);
    }

    find(x) {
        if(this.parent[x] !== x) {
            this.parent[x] = find(this.parent[x]);
        }
        return this.parent[x];
    }

    uniion(x, y) {
        let x = find(x);
        let y = find(y);
        console.log(x, y);
    }
}